test/java/org/chromium/net/impl/CancelProofEnvoyStreamTest.java

```diff
@@ -0,0 +1,379 @@
+    mSendDataBlock.close(); // Following Thread will block on executing sendData
+    new Thread(() -> cancelProofEnvoyStream.sendData(BYTE_BUFFER, false)).start();
+    mStartLatch = new CountDownLatch(1); // Only one method to wait for.
+    mStartLatch.await();                 // Wait for the above Thread to enter the sendData method.
+
+    cancelProofEnvoyStream.cancel();
+
+    assertThat(mSendDataInvocationCount.get()).isOne();
+    assertThat(mCancelInvocationCount.get()).isZero();
+  }
+
+  @Test
+  public void cancel_whileReadDataIsExecuting_cancelGetsPostponed() throws Exception {
+    cancelProofEnvoyStream.setStream(mMockedStream);
+    mReadDataBlock.close(); // Following Thread will block on executing readData
+    new Thread(() -> cancelProofEnvoyStream.readData(1)).start();
+    mStartLatch = new CountDownLatch(1); // Only one method to wait for.
+    mStartLatch.await();                 // Wait for the above Thread to enter the readData method.
+
+    cancelProofEnvoyStream.cancel();
+
+    assertThat(mReadDataInvocationCount.get()).isOne();
+    assertThat(mCancelInvocationCount.get()).isZero();
+  }
+
+  @Test
+  public void cancel_manyConcurrentStreamOperationsInFlight() throws Exception {
+    cancelProofEnvoyStream.setStream(mMockedStream);
+    Thread[] threads = new Thread[300];
+    AtomicInteger count = new AtomicInteger();
+    for (int i = 0; i < threads.length; i++) {
+      threads[i] = new Thread(() -> {
+        switch (count.getAndIncrement() % 3) {
+        case 0:
+          cancelProofEnvoyStream.sendHeaders(HEADERS, false);
+          break;
+        case 1:
+          cancelProofEnvoyStream.sendData(BYTE_BUFFER, false);
+          break;
+        case 2:
+          cancelProofEnvoyStream.readData(1);
+          break;
+        }
+      });
+    }
+    // Every Thread will be blocked while executing one of the cancelProofEnvoyStream's method.
+    mSendHeadersBlock.close();
+    mSendDataBlock.close();
+    mReadDataBlock.close();
+    mStartLatch = new CountDownLatch(threads.length);
+    for (Thread thread : threads) {
+      thread.start();
+    }
+    mStartLatch.await(); // Wait for all Thread to reach the blocking point.
+    // Sanity check
+    assertThat(mSendHeadersInvocationCount.get()).isEqualTo(threads.length / 3);
+    assertThat(mSendDataInvocationCount.get()).isEqualTo(threads.length / 3);
+    assertThat(mReadDataInvocationCount.get()).isEqualTo(threads.length / 3);
+
+    cancelProofEnvoyStream.cancel();
+    assertThat(mCancelInvocationCount.get()).isZero();
+
+    // Let all the Threads to finish executing the cancelProofEnvoyStream's methods.
+    mSendHeadersBlock.open();
+    mSendDataBlock.open();
+    mReadDataBlock.open();
+    // Wait for all the Threads to die.
+    for (Thread thread : threads) {
+      thread.join();
+    }
+    assertThat(mCancelInvocationCount.get()).isOne();
+  }
+
+  private class MockedStream extends EnvoyHTTPStream {
```

Comment:
Can you add a comment which explains the various Block/Latch behaviors of this class?

Reply:
Added a comment. Short story: a `ConditionVariable` is in reality a `CountDownLatch` where the initial count is "1".